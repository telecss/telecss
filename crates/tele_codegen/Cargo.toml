[package]
name = "tele_codegen"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tele_parser = { path = "../tele_parser" }
