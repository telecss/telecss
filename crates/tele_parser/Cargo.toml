[package]
name = "tele_parser"
version = "0.2.0"
description = "A CSS parser"
edition = "2021"
license = "MIT"
repository = "https://github.com/telecss/telecss"
keywords = ["css", "telecss", "tele_parser", "parser"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tele_tokenizer = { path = "../tele_tokenizer", version = "0.2.0" }
serde = { version = "1.0.133", features = ["derive", "rc"] }

[dev-dependencies]
insta = "1.10.0"