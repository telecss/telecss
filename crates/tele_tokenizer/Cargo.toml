[package]
name = "tele_tokenizer"
version = "0.1.0"
description = "A CSS tokenizer"
edition = "2021"
repository = "https://github.com/telecss/telecss"
license = "MIT"
keywords = ["css", "telecss", "tele_tokenizer", "tokenizer"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tele_utils = { path = "../tele_utils", version = "0.1.0" }
serde = { version = "1.0.133", features = ["derive"] }

[dev-dependencies]
insta = "1.10.0"

[features]
