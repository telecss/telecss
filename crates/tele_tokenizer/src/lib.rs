#![warn(missing_docs)]
#![deny(unused_imports)]
#![deny(unused_variables)]

//! CSS Tokenizer

use std::{
  iter::{Enumerate, Peekable},
  slice::Iter,
  str::from_utf8_unchecked,
  vec,
};

mod error;
mod inner_state;
mod token;

use inner_state::*;
use tele_utils::*;

pub use error::*;
pub use token::*;

#[derive(Debug)]
/// Represents the CSS tokenizer.
pub struct Tokenizer<'s> {
  bytes: &'s [u8],
  iter: Peekable<Enumerate<Iter<'s, u8>>>,
  line: usize,
  colnmu: usize,
  offset: usize,
  pre_cursor: Pos,
  current_state: State,
  tokens: Vec<Token<'s>>,
}

impl<'s> From<&'s str> for Tokenizer<'s> {
  fn from(source: &'s str) -> Self {
    Tokenizer {
      bytes: source.as_bytes(),
      iter: source.as_bytes().iter().enumerate().peekable(),
      // position related
      line: 1,
      colnmu: 1,
      offset: 0,
      pre_cursor: Pos::default(),
      current_state: State::Initial,
      tokens: Vec::new(),
    }
  }
}

impl<'s> Tokenizer<'s> {
  /// Perform tokenizing
  pub fn tokenize(&mut self) -> Result<&Vec<Token<'s>>> {
    while let Some(&(offset, &c)) = self.iter.peek() {
      let c = c as char;
      self.current_state = match self.current_state {
        State::Initial => match c {
          c if is_whitespace(c) => State::WhiteSpace,
          c if is_ident_start(c) => State::IdentLike,
          c if is_digit(c) => State::Numeric,
          '.' => {
            // https://www.w3.org/TR/css-syntax-3/#starts-with-a-number
            if is_start_a_number(&self.bytes[offset..]) {
              State::Numeric
            } else {
              self.consume(offset, 1, false);
              self.emit(TokenType::Delim);
              State::Initial
            }
          }
          '"' | '\'' => State::String,
          '{' => {
            self.consume(offset, 1, false);
            self.emit(TokenType::LeftCurlyBracket);
            State::Initial
          }
          '}' => {
            self.consume(offset, 1, false);
            self.emit(TokenType::RightCurlyBracket);
            State::Initial
          }
          ')' => {
            self.consume(offset, 1, false);
            self.emit(TokenType::RightParentheses);
            State::Initial
          }
          '(' => {
            self.consume(offset, 1, false);
            self.emit(TokenType::LeftParentheses);
            State::Initial
          }
          '[' => {
            self.consume(offset, 1, false);
            self.emit(TokenType::LeftSquareBracket);
            State::Initial
          }
          ']' => {
            self.consume(offset, 1, false);
            self.emit(TokenType::RightSquareBracket);
            State::Initial
          }
          ':' => {
            self.consume(offset, 1, false);
            self.emit(TokenType::Colon);
            State::Initial
          }
          ';' => {
            self.consume(offset, 1, false);
            self.emit(TokenType::SemiColon);
            State::Initial
          }
          '/' if is_comment_start(&self.bytes[offset..]) => {
            self.consume(offset, 2, false);
            State::Comment
          }
          ',' => {
            self.consume(offset, 1, false);
            self.emit(TokenType::Comma);
            State::Initial
          }
          '#' => {
            let cp1 = char_at(&self.bytes[offset..], 1);

            if is_ident_char(cp1) {
              self.consume(offset, 1, true); // consume #
              self.consume_ident_seq();
              self.emit(TokenType::Hash);
            } else {
              self.consume(offset, 1, false);
              self.emit(TokenType::Delim);
            }
            State::Initial
          }
          '+' => {
            if is_start_a_number(&self.bytes[offset..]) {
              self.consume_a_numeric();
            } else {
              self.consume(offset, 1, false);
              self.emit(TokenType::Delim);
            }
            State::Initial
          }
          '-' => {
            let mut returned_state = State::Initial;
            let cp1 = char_at(&self.bytes[offset..], 1);
            let cp2 = char_at(&self.bytes[offset..], 2);
            if is_start_a_number(&self.bytes[offset..]) {
              self.consume_a_numeric();
            } else if cp1 == '-' && cp2 == '>' {
              self.consume(offset, 3, false);
              self.emit(TokenType::CDC);
            } else if would_start_an_ident_seq(&self.bytes[offset..]) {
              returned_state = State::IdentLike;
            } else {
              self.consume(offset, 1, false);
              self.emit(TokenType::Delim);
            }
            returned_state
          }
          '<' => {
            let cp1 = char_at(&self.bytes[offset..], 1);
            let cp2 = char_at(&self.bytes[offset..], 2);
            let cp3 = char_at(&self.bytes[offset..], 3);

            if cp1 == '!' && cp2 == '-' && cp3 == '-' {
              self.consume(offset, 4, false);
              self.emit(TokenType::CDO);
            } else {
              self.consume(offset, 1, false);
              self.emit(TokenType::Delim);
            }
            State::Initial
          }
          '@' => {
            if would_start_an_ident_seq(&self.bytes[(offset + 1)..]) {
              self.consume(offset, 1, true); // ignore @
              self.consume_ident_seq();
              self.emit(TokenType::AtKeyword);
            } else {
              self.consume(offset, 1, false);
              self.emit(TokenType::Delim);
            }
            State::Initial
          }
          '\\' => {
            if is_valid_escape(&self.bytes[offset..]) {
              State::IdentLike
            } else {
              self.consume(offset, 1, false);
              self.emit(TokenType::Delim);
              State::Initial
            }
          }
          _ => {
            self.consume(offset, 1, false);
            // anything else, return a <delim-token>
            self.emit(TokenType::Delim);
            State::Initial
          }
        },
        State::WhiteSpace => {
          if is_whitespace(c) {
            if is_newline(c) {
              self.line += 1;
              self.colnmu = 1;
            }
            self.consume(offset, 1, false);
            State::WhiteSpace
          } else {
            self.emit(TokenType::WhiteSpace);
            State::Initial
          }
        }
        // https://www.w3.org/TR/css-syntax-3/#consume-an-ident-like-token
        State::IdentLike => {
          if would_start_an_ident_seq(&self.bytes[offset..]) {
            let (start_pos, end_pos) = self.consume_ident_seq();
            let ident = unsafe {
              from_utf8_unchecked(&self.bytes[start_pos.offset..end_pos.offset])
                .to_ascii_lowercase()
            };

            let next = char_at(&self.bytes[end_pos.offset..], 0);
            let mut returned_state = State::Initial;

            if ident == "url" && next == '(' {
              self.emit(TokenType::URL);
              // consume '('
              self.consume(end_pos.offset, 1, false);
              self.emit(TokenType::LeftParentheses);

              let (start_pos, end_pos, is_function_token) = self.consume_whitespace_for_url_call();

              // should emit whitespace token when is_function_token is true
              let should_emit_ws = is_function_token && start_pos.offset < end_pos.offset;
              if should_emit_ws {
                self.emit(TokenType::WhiteSpace);
              }
              if is_function_token {
                // change URL token to Function token
                self
                  .tokens
                  .iter_mut()
                  .rev()
                  .skip_while(|ty| ty.is_ws() || ty.is_lp())
                  .next()
                  .map(|token| token.token_type = TokenType::Function);
              } else {
                if should_emit_ws {
                  self.tokens.pop();
                }
                self.tokens.pop(); // remove TokenType::LeftParentheses
                self.tokens.pop(); // remove TokenType::URL
                returned_state = State::URL;
              }
            } else if next == '(' {
              self.emit(TokenType::Function);
              self.consume(end_pos.offset, 1, false);
              self.emit(TokenType::LeftParentheses);
            } else {
              self.emit(TokenType::Ident);
            }
            returned_state
          } else {
            return Err(Error::from(ErrorKind::InvalidIdentSeq));
          }
        }
        State::URL => {
          // https://www.w3.org/TR/css-syntax-3/#consume-a-url-token
          self.consume_url_token()?;
          State::Initial
        }
        // https://www.w3.org/TR/css-syntax-3/#consume-a-string-token
        State::String => {
          let ending_char = c;
          let mut valid_string_token: bool = false;
          self.consume(offset, 1, true);
          while let Some(&(offset, &c)) = self.iter.peek() {
            let c = c as char;
            let next = char_at(&self.bytes[offset..], 1);
            if is_newline(c) {
              return Err(Error::from(ErrorKind::BadString));
            } else if c == ending_char {
              self.emit(TokenType::String);
              // ignore ending_char
              self.consume(offset, 1, true);
              valid_string_token = true;
              break;
            } else if c == '\\' && next == ending_char {
              self.consume(offset, 2, false);
            } else {
              self.consume(offset, 1, false);
            }
          }
          if !valid_string_token {
            return Err(Error::from(ErrorKind::BadString));
          }
          State::Initial
        }
        // https://www.w3.org/TR/css-syntax-3/#consume-comment
        State::Comment => {
          if is_comment_end(&self.bytes[offset..]) {
            self.consume(offset, 2, false);
            self.emit(TokenType::Comment);
            State::Initial
          } else {
            self.consume(offset, 1, false);
            State::Comment
          }
        }
        State::Numeric => {
          self.consume_a_numeric();
          State::Initial
        }
        State::EOF => break,
      };
    }

    // EOF
    match self.current_state {
      State::WhiteSpace => {
        self.emit(TokenType::WhiteSpace);
      }
      State::Comment | State::URL => return Err(Error::from(ErrorKind::UnexpectedEOF)),
      _ => {}
    }
    self.emit(TokenType::EOF);

    Ok(&mut self.tokens)
  }

  // https://www.w3.org/TR/css-syntax-3/#consume-numeric-token
  fn consume_a_numeric(&mut self) {
    let (_, end_pos) = self.consume_a_number();
    let next = char_at(&self.bytes[end_pos.offset..], 0);

    if would_start_an_ident_seq(&self.bytes[end_pos.offset..]) {
      // According to the spec: https://www.w3.org/TR/css-syntax-3/#consume-numeric-token,
      // we need to continue to analyze the unit of the Die token,
      // but we can also choose to analyze the unit as a separate `Ident` token
      // e.g.
      // `30px` -> TokenType::Dimension + TokenType::Ident
      //                    (30)                   (px)
      self.emit(TokenType::Dimension);
    } else if next == '%' {
      self.emit(TokenType::Percentage);
      self.consume(end_pos.offset, 1, true);
    } else {
      self.emit(TokenType::Number);
    }
  }

  // https://www.w3.org/TR/css-syntax-3/#consume-a-number
  fn consume_a_number(&mut self) -> (Pos, Pos) {
    let start_pos = self.get_cursor();
    match self.iter.peek() {
      Some(&(offset, &b'+')) | Some(&(offset, &b'-')) => {
        self.consume(offset, 1, false);
      }
      _ => {}
    }

    self.try_consume_digit();

    if let Some(&(offset, _)) = self.iter.peek() {
      let cp1 = char_at(&self.bytes[offset..], 0);
      let cp2 = char_at(&self.bytes[offset..], 1);

      if cp1 == '.' && is_digit(cp2) {
        self.consume(offset, 2, false);
        self.try_consume_digit();
      }
    }

    if let Some(&(offset, _)) = self.iter.peek() {
      let cp1 = char_at(&self.bytes[offset..], 0);
      let cp2 = char_at(&self.bytes[offset..], 1);
      let cp3 = char_at(&self.bytes[offset..], 2);

      if cp1 == 'E' || cp1 == 'e' {
        if (cp2 == '+' || cp2 == '-') && is_digit(cp3) {
          self.consume(offset, 3, false);
          self.try_consume_digit();
        } else if is_digit(cp2) {
          self.consume(offset, 2, false);
          self.try_consume_digit();
        }
      }
    }
    let end_pos = self.get_cursor();

    (start_pos, end_pos)
  }

  fn try_consume_digit(&mut self) {
    while let Some(&(offset, &c)) = self.iter.peek() {
      if is_digit(c as char) {
        self.consume(offset, 1, false);
      } else {
        break;
      }
    }
  }

  fn consume_ident_seq(&mut self) -> (Pos, Pos) {
    let start_cursor = self.get_cursor();
    while let Some(&(offset, &c)) = self.iter.peek() {
      if is_ident_char(c as char) {
        self.consume(offset, 1, false);
      } else {
        break;
      }
    }
    let end_cursor = self.get_cursor();
    (start_cursor, end_cursor)
  }

  fn consume_whitespace_for_url_call(&mut self) -> (Pos, Pos, bool) {
    let start_cursor = self.get_cursor();
    let mut is_function_token = false;
    while let Some(&(offset, &c)) = self.iter.peek() {
      let c = c as char;
      if is_whitespace(c) {
        self.consume(offset, 1, false);
      } else if c == '"' || c == '\'' {
        is_function_token = true;
        break;
      } else {
        break;
      }
    }
    let end_cursor = self.get_cursor();
    (start_cursor, end_cursor, is_function_token)
  }

  fn consume_url_token(&mut self) -> Result<()> {
    while let Some(&(offset, &c)) = self.iter.peek() {
      let c = c as char;
      match c {
        ')' => {
          self.emit(TokenType::URL);
          break;
        }
        '"' | '\'' | '(' => return Err(Error::from(ErrorKind::BadURL)),
        c if is_non_printable(c) => return Err(Error::from(ErrorKind::BadURL)),
        c if is_whitespace(c) => {
          self.emit(TokenType::URL);
          let (_, end_pos, _) = self.consume_whitespace_for_url_call();
          let next = char_at(&self.bytes[end_pos.offset..], 0);
          if next != ')' {
            self.tokens.pop();
            return Err(Error::from(ErrorKind::BadURL));
          } else {
            self.consume(end_pos.offset, 1, true);
            break;
          }
        }
        _ => {
          self.consume(offset, 1, false);
        }
      }
    }
    Ok(())
  }

  fn consume(&mut self, offset: usize, count: usize, advance_cursor: bool) -> Vec<u8> {
    let mut buf: Vec<u8> = vec![];
    for n in 0..count {
      if let Some((_, c)) = self.iter.next() {
        buf.push(*c);
      }
      self.colnmu += 1;
      self.offset = offset + n;
    }
    if advance_cursor {
      self.pre_cursor = self.get_cursor();
    }

    buf
  }

  fn get_cursor(&self) -> Pos {
    Pos {
      offset: if self.colnmu == 1 { 0 } else { self.offset + 1 },
      line: self.line,
      column: self.colnmu,
    }
  }

  fn emit(&mut self, token_type: TokenType) -> &mut Token<'s> {
    let cur_cursor = self.get_cursor();

    let token = Token::new(
      token_type,
      self.pre_cursor,
      cur_cursor,
      &self.bytes[self.pre_cursor.offset..cur_cursor.offset],
    );

    self.pre_cursor = cur_cursor;
    self.tokens.push(token);

    self.tokens.last_mut().unwrap()
  }
}
