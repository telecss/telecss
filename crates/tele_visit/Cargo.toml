[package]
name = "tele_visit"
version = "0.1.0"
description = "A CSS AST visitor for visit the AST produced by tele_parser"
edition = "2021"
repository = "https://github.com/telecss/telecss"
license = "MIT"
keywords = ["css", "telecss", "tele_visit", "visit"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tele_parser = { path = "../tele_parser", version = "0.1.0" }

[dev-dependencies]
insta = "1.10.0"
tele_tokenizer = { path = "../tele_tokenizer", version = "0.1.0" }