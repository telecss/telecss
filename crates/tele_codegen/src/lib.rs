#![warn(missing_docs)]
#![deny(unused_imports)]
#![deny(unused_variables)]

//! A CSS code generator that generates CSS code from the AST generated by tele_parser

use std::error::Error;
use std::fmt::Write;
use tele_parser::{
  AstType, AtRuleNode, DeclarationNode, RuleSetNode, StatementNode, StyleSheetNode,
};

pub trait CodeGenerator {
  fn generate(&mut self, ast: AstType) -> Result<(), Box<dyn Error>> {
    self.gen_ss_node(&*ast.borrow())
  }

  fn gen_ss_node(&mut self, ss_node: &StyleSheetNode) -> Result<(), Box<dyn Error>>;
  fn gen_rule_set_node(&mut self, rule_set_node: &RuleSetNode) -> Result<(), Box<dyn Error>>;
  fn gen_at_rule_node(&mut self, at_rule_node: &AtRuleNode) -> Result<(), Box<dyn Error>>;
  fn gen_decl_node(&mut self, decl_node: &DeclarationNode) -> Result<(), Box<dyn Error>>;

  fn gen_statements(&mut self, statements: &Vec<StatementNode>) -> Result<(), Box<dyn Error>> {
    for stat in statements {
      match stat {
        StatementNode::AtRule(node) => self.gen_at_rule_node(&*node.borrow_mut())?,
        StatementNode::RuleSet(node) => self.gen_rule_set_node(&*node.borrow_mut())?,
      }
    }

    Ok(())
  }
}

pub struct Codegen<'a, W: Write> {
  css: &'a mut W,
  line: usize,
  column: usize,
  indent: usize,
  indent_level: usize,
}

impl<'a, W: Write> Codegen<'a, W> {
  pub fn new(writer: &'a mut W) -> Self {
    Codegen {
      css: writer,
      line: 0,
      column: 1,
      indent: 2,
      indent_level: 0,
    }
  }

  fn new_line(&mut self) -> Result<(), Box<dyn Error>> {
    self.line += 1;
    self.css.write_char('\n')?;
    for _ in 0..self.indent_level * self.indent {
      self.css.write_char(' ')?;
    }

    Ok(())
  }

  fn indent(&mut self) {
    self.indent_level += 1;
  }

  fn de_indent(&mut self) {
    self.indent_level -= 1;
  }
}

impl<'a, W: Write> CodeGenerator for Codegen<'a, W> {
  fn gen_ss_node(&mut self, ss_node: &StyleSheetNode) -> Result<(), Box<dyn Error>> {
    self.gen_statements(&ss_node.statements)
  }
  fn gen_rule_set_node(&mut self, rule_set_node: &RuleSetNode) -> Result<(), Box<dyn Error>> {
    self.css.write_str(&rule_set_node.prelude)?;
    self.css.write_str(" {")?;
    self.indent();

    for decl in &rule_set_node.declarations {
      self.new_line()?;
      self.gen_decl_node(&*decl.borrow())?;
    }

    self.de_indent();
    self.new_line()?;
    self.css.write_char('}')?;

    Ok(())
  }

  fn gen_at_rule_node(&mut self, at_rule_node: &AtRuleNode) -> Result<(), Box<dyn Error>> {
    self.css.write_char('@')?;
    self.css.write_str(&at_rule_node.name)?;
    self.css.write_char(' ')?;
    self.css.write_str(&at_rule_node.prelude)?;

    if at_rule_node.block.len() > 0 {
      self.css.write_char('{')?;
      self.indent();
      self.new_line()?;
      self.gen_statements(&at_rule_node.block)?;
      self.de_indent();
      self.new_line()?;
      self.css.write_char('}')?;
    } else {
      self.css.write_char(';')?;
    }

    Ok(())
  }

  fn gen_decl_node(&mut self, decl_node: &DeclarationNode) -> Result<(), Box<dyn Error>> {
    self.css.write_str(&decl_node.name)?;
    self.css.write_str(": ")?;
    self.css.write_str(&decl_node.value)?;

    if decl_node.important {
      self.css.write_str(" !important;")?;
    } else {
      self.css.write_char(';')?;
    }

    Ok(())
  }
}
