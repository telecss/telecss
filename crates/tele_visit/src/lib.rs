#![warn(missing_docs)]
#![deny(unused_imports)]
#![deny(unused_variables)]

//! Visitor/Folder pattern for transforming the AST

use tele_parser::{AtRuleNode, DeclarationNode, RuleSetNode, StatementNode, StyleSheetNode};

/// [VisitMut] can modify AST nodes *in place*
pub trait VisitMut<'s> {
  /// Visit [StyleSheetNode] node
  fn visit_ss_node(&mut self, mut ss_node: StyleSheetNode<'s>) -> StyleSheetNode<'s> {
    ss_node.statements = ss_node
      .statements
      .into_iter()
      .map(|stat| match stat {
        StatementNode::AtRule(node) => StatementNode::AtRule(self.visit_at_rule_node(node)),
        StatementNode::RuleSet(node) => StatementNode::RuleSet(self.visit_rule_set_node(node)),
      })
      .collect();

    ss_node
  }

  /// Visit [RuleSetNode] node
  fn visit_rule_set_node(&mut self, mut rule_set_node: RuleSetNode<'s>) -> RuleSetNode<'s> {
    rule_set_node.declarations = rule_set_node
      .declarations
      .into_iter()
      .map(|decl| self.visit_decl_node(decl))
      .collect();

    rule_set_node
  }

  /// Visit [AtRuleNode] node
  fn visit_at_rule_node(&mut self, mut at_rule_node: AtRuleNode<'s>) -> AtRuleNode<'s> {
    at_rule_node.block = at_rule_node
      .block
      .into_iter()
      .map(|stat| match stat {
        StatementNode::AtRule(node) => StatementNode::AtRule(self.visit_at_rule_node(node)),
        StatementNode::RuleSet(node) => StatementNode::RuleSet(self.visit_rule_set_node(node)),
      })
      .collect();

    at_rule_node
  }

  /// Visit [DeclarationNode] node
  fn visit_decl_node(&mut self, decl_node: DeclarationNode<'s>) -> DeclarationNode<'s> {
    decl_node
  }
}

/// Transform the AST according to the given visitor.
pub fn transform<'a, T: VisitMut<'a>>(
  ast: StyleSheetNode<'a>,
  mut visitor: T,
) -> StyleSheetNode<'a> {
  visitor.visit_ss_node(ast)
}
