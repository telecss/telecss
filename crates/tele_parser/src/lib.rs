#![warn(missing_docs)]
#![deny(unused_imports)]
#![deny(unused_variables)]

//! A simpler and faster CSS parser.
//!
//! # Usages
//! ```rust
//! use tele_tokenizer::Tokenizer;
//! use tele_parser::Parser;
//!
//! // Create a tokenizer
//! let mut tokenizer: Tokenizer = r".foo { color: red; }".into();
//! // Tokenize based on the given raw string,
//! let tokens = tokenizer.tokenize().unwrap();
//! // Create a parser from a sequence of tokens
//! let parser = Parser::from(tokens);
//! // parsing it
//! let ast = parser.parse();
//! ```

mod error;
mod node;

pub use error::Result as ParserResult;
use error::{SyntaxError, SyntaxErrorKind};
pub use node::*;

use std::{cell::RefCell, iter::Peekable, rc::Rc, slice::Iter};

use tele_tokenizer::*;

/// Represents the CSS parser.
pub struct Parser<'s> {
  iter: RefCell<Peekable<Iter<'s, Token<'s>>>>,
}

impl<'s> From<&'s Vec<Token<'s>>> for Parser<'s> {
  fn from(tokens: &'s Vec<Token<'s>>) -> Self {
    Self {
      iter: RefCell::new(tokens.iter().peekable()),
    }
  }
}

impl<'s> Parser<'s> {
  /// Perform parsing
  pub fn parse(&self) -> ParserResult<Rc<RefCell<StyleSheetNode>>> {
    let mut ss_node = StyleSheetNode::default();

    while let Some(node) = self.parse_statements()? {
      ss_node.statements.push(node)
    }

    self.skip_ws_and_comments();

    let next = self.peek();
    debug_assert!(next.is_eof());

    ss_node.loc.end = next.end_pos;

    Ok(Rc::new(RefCell::new(ss_node)))
  }

  fn peek(&self) -> &Token {
    let next = self.iter.borrow_mut().peek().map(|&token| token);
    // Because the last token will always be EOF2, and we will handle it reasonably.
    // So in general/theoretically, we don't encounter `None`.
    debug_assert!(next.is_some());
    unsafe { next.unwrap_unchecked() }
  }

  fn consume(&self) -> Option<&Token> {
    self.iter.borrow_mut().next()
  }

  fn parse_statements(&self) -> ParserResult<Option<StatementNode>> {
    self.skip_ws_and_comments();

    let next = self.peek();

    if next.is_eof() {
      return Ok(None);
    }

    if next.is_at_keyword() {
      Ok(
        self
          .parse_at_rule()?
          .map(|node| StatementNode::AtRule(Rc::new(RefCell::new(node)))),
      )
    } else {
      Ok(
        self
          .parse_rule_set()?
          .map(|node| StatementNode::RuleSet(Rc::new(RefCell::new(node)))),
      )
    }
  }

  fn parse_rule_set(&self) -> ParserResult<Option<RuleSetNode>> {
    let mut rule_set_node = RuleSetNode::default();

    let mut token = self.peek();
    rule_set_node.loc.start = token.start_pos;

    loop {
      token = self.peek();
      self.asset_eof(token)?;

      if token.is_lcb() {
        self.consume();
        while let Some(node) = self.parse_decl()? {
          rule_set_node.declarations.push(Rc::new(RefCell::new(node)));
        }
      } else if token.is_rcb() {
        self.consume();
        break;
      } else {
        if !token.is_comment() {
          rule_set_node.prelude.push_str(&token.to_string());
          rule_set_node.prelude_tokens.push(token);
        }
        self.consume();
      }
    }

    rule_set_node.prelude = rule_set_node.prelude.trim_end().to_string();
    rule_set_node.loc.end = token.end_pos;

    Ok(Some(rule_set_node))
  }

  fn parse_at_rule(&self) -> ParserResult<Option<AtRuleNode>> {
    let mut at_rule_node = AtRuleNode::default();

    // unsafe: we are sure that the next token is `@`
    let mut token = unsafe { self.consume().unwrap_unchecked() };
    at_rule_node.loc.start = token.start_pos;
    at_rule_node.name = token.to_string();
    at_rule_node.name_tokens.push(token);

    self.skip_ws_and_comments();

    loop {
      token = self.peek();
      self.asset_eof(token)?;

      if token.is_lcb() {
        self.consume();

        loop {
          self.skip_ws_and_comments();
          let next = self.peek();
          if next.is_rcb() {
            // end right curly bracket
            self.consume();
            break;
          }

          let statement = self.parse_statements()?;
          if statement.is_none() {
            break;
          }

          // unsafe: checked above
          let statement = unsafe { statement.unwrap_unchecked() };

          at_rule_node.block.push(statement);
        }
        break;
      } else if token.is_semi_colon() {
        // no-block
        self.consume();
        break;
      } else {
        // prelude
        self.consume();
        at_rule_node.prelude.push_str(&token.to_string());
        at_rule_node.prelude_tokens.push(token);
      }
    }

    at_rule_node.loc.end = token.end_pos;
    at_rule_node.prelude = at_rule_node.prelude.trim_end().to_string();

    Ok(Some(at_rule_node))
  }

  fn parse_decl(&self) -> ParserResult<Option<DeclarationNode>> {
    self.skip_ws_and_comments();

    // declaration name

    let next = self.peek();
    if next.is_rcb() {
      return Ok(None);
    }
    self.asset_eof(next)?;

    if !next.is_ident() {
      return Err(Box::new(SyntaxError::from(
        SyntaxErrorKind::InvalidDeclName,
      )));
    }

    let mut decl_node = DeclarationNode::default();
    decl_node.name = next.to_string();
    decl_node.name_tokens.push(next);
    decl_node.loc.start = next.start_pos;

    self.consume();
    self.skip_ws_and_comments();

    // :

    let next = self.peek();
    if !next.is_colon() {
      return Err(Box::new(SyntaxError::from(SyntaxErrorKind::MissingColon)));
    }
    self.asset_eof(next)?;
    self.consume();
    self.skip_ws_and_comments();

    // declaration value

    let mut last_one: Option<&Token> = None;
    let mut second_last: Option<&Token> = None;

    loop {
      let token = self.peek();
      if token.is_eof() {
        self.asset_eof(token)?;
      } else if token.is_semi_colon() {
        self.consume();
        break;
      } else if token.is_rcb() {
        decl_node.value = decl_node.value.trim_end().to_string();
        break;
      }

      self.consume();

      if !token.is_ws() && !token.is_comment() {
        if last_one.is_some() {
          second_last = last_one;
        }
        last_one = Some(token);
      }

      if !token.is_comment() {
        decl_node.value.push_str(&token.to_string());
        decl_node.value_tokens.push(token);
      }
    }

    decl_node.important = second_last.is_some()
      && unsafe { second_last.unwrap_unchecked() }.is_delim(&[b'!'])
      && last_one.is_some()
      && unsafe { last_one.unwrap_unchecked() }.is_ident_with(b"important");

    if decl_node.important {
      // remove the trailing `!important`
      decl_node.value = decl_node.value[..decl_node.value.as_str().len() - 11]
        // trim whitespaces between the `value` and `!important`
        .trim_end()
        .to_string();
    }

    last_one.map(|token| decl_node.loc.end = token.end_pos);

    Ok(Some(decl_node))
  }

  fn asset_eof(&self, token: &Token) -> ParserResult<()> {
    if token.is_eof() {
      return Err(Box::new(SyntaxError::from(SyntaxErrorKind::UnexpectedEOF)));
    }
    Ok(())
  }

  fn skip_ws_and_comments(&self) {
    loop {
      let token = self.peek();
      if token.is_ws() || token.is_comment() {
        self.consume();
      } else {
        break;
      }
    }
  }
}
