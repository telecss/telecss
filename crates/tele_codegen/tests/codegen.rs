#![allow(unused_must_use)]

use insta::assert_snapshot;
use std::{cell::RefCell, rc::Rc};
use tele_codegen::{CodeGenerator, Codegen};
use tele_parser::*;
use tele_tokenizer::*;
use tele_visit::*;

struct Renamer;
impl<'s> MutVisitor<'s> for Renamer {
  fn visit_decl_node(&self, decl_node: Rc<RefCell<DeclarationNode<'s>>>) {
    let mut decl_node = decl_node.borrow_mut();
    decl_node.name = format!("-moz-{}", decl_node.name);
  }
}

#[test]
fn codegen_for_ruleset() {
  // tokenizing
  let mut tokenizer: Tokenizer = r"  .foo { color: red; }  ".into();
  let tokens = tokenizer.tokenize().unwrap();
  // parsing
  let parser = Parser::from(tokens);
  let ast = parser.parse().unwrap();
  // transforming
  let transformer = VisitMut::new(Rc::clone(&ast), vec![Box::new(Renamer)]);
  transformer.transform();

  // codegen
  let mut css = String::new();
  let mut codegen = Codegen::new(&mut css);
  codegen.generate(ast);

  assert_snapshot!(&css);
}

#[test]
fn codegen_for_at_rule() {
  // tokenizing
  let mut tokenizer: Tokenizer = r"
  @media screen and (min-width: 900px) { article { padding: 1rem 3rem; } }"
    .into();
  let tokens = tokenizer.tokenize().unwrap();
  // parsing
  let parser = Parser::from(tokens);
  let ast = parser.parse().unwrap();
  // transforming
  let transformer = VisitMut::new(Rc::clone(&ast), vec![Box::new(Renamer)]);
  transformer.transform();

  // codegen
  let mut css = String::new();
  let mut codegen = Codegen::new(&mut css);
  codegen.generate(ast);

  assert_snapshot!(&css);
}
