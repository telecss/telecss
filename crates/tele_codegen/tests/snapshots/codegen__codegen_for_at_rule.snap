---
source: crates/tele_codegen/tests/codegen.rs
assertion_line: 57
expression: "&css"

---
@media screen and (min-width: 900px){
  article {
    -moz-padding: 1rem 3rem;
  }
}
